Fix issues in Cryptocell 310 shax_alt discovered by On Target Testing

RonEld opened the pull request:
### Description

Initialize the Cryptocell context in the `mbedtls_shax_process()`
function, in case it wasn't initialized. The Process function can be
called without calling to the starts function. Discovered by
On Target Testing on the NRF52840_DK platform.

Note: This might have conflicts with #8643 , but not dependent on it.
<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

0xc0170 commented:
@RonEld is @ARMmbed/mbed-os-tls the team that should review this fix? If yes, please review

Commit:
Style fixes

1. Change camelcase variables to Mbed OS style.
2. Remove functions declarations from the `_alt` header, since they
are now added from the module header regardless whether an alternative
implementation exists.
3. Remove the `extern "c"` declaration from the `_alt` headers.
4. Remove redundant extra lines.

RonEld commented:
@0xc0170 I would say the @ARMmbed/mbed-os-crypto team should review this

cmonr commented:
Restarting CI

0xc0170 commented:
Aborting CI, will start once it is approved

0xc0170 commented:
It was completed, waiting for reviewers

yanesca commented on features/cryptocell/FEATURE_CRYPTOCELL310/sha1_alt.c:
```diff
@@ -79,8 +91,21 @@ int mbedtls_sha1_finish_ret( mbedtls_sha1_context *ctx,
+    int ret = 0;
+    if( ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( ctx );
```
I don't think that calling `mbedtls_internal_sha1_process()` should be supported for several reasons:
1. This is an internal function, users shouldn't call it anyway.
2. If anybody, Mbed TLS, the tests or our users call it without calling `mbedtls_sha1_starts()` first, then they are violating the API

RonEld replied:
As you can see::
[`sha1_process_wrap`](https://github.com/ARMmbed/mbedtls/blob/development/library/md_wrap.c#L366) calls `mbedtls_internal_sha1_process()`. I guess it's ok, since it's internal. The test for [`mbedtlos_md_process()`](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_md.function#L11) only calls setup and process, so perhaps it's a bug in the test...

yanesca replied:
I think it is a bug in the test, it should call `mbedtls_md_starts( )` too.

RonEld replied:
created https://github.com/ARMmbed/mbedtls/issues/2227 to track this

yanesca commented on features/cryptocell/FEATURE_CRYPTOCELL310/sha1_alt.h:
```diff
@@ -33,6 +33,7 @@ extern "C" {
 typedef struct
 {
     CRYS_HASHUserContext_t crys_hash_ctx;
+    int                    is_cc_initiated;
```
style: I think the word we are looking for here is "initialised" and not  "initiated".

RonEld replied:
OK.

yanesca reviewed (changes requested):
I don't think that this is the right way to resolve the issue. I left a comment about my thoughts in the code.

RonEld commented:
@yanesca Thanks for the review!
If we agreee the issue is in the test code, and not in the driver, then I guess we can close this PR

yanesca commented:
I think we agree, but I think that the style changes in this PR are still valuable. Maybe we could still merge those or just cherry-pick them to the top of one of the other PRs.

Also, before doing anything we should raise an issue for the bug in the tests. Which repository are they in?

RonEld commented:
The tests are from the Mbed TLS repository
I can change do the style changes early next week

yanesca commented:
Can you point me to the failing test and/or raise an issue about this in the Mbed TLS repository?

RonEld commented:
@yanesca [`mbedtls_md_process()`](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_md.function#L11) is the failing test.
However, after thinking some more, I think we still need to have this initiated mechanism.
The TLS code, for example [here](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L1135) clones an already started SHA context. I think that once cloning, we should consider the context as not initiated, as the HW may have switched context, so we need to initiate again.

Commit:
Return not supported for internal process function

Return the `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` for the
internal process function, as it is not public.

RonEld commented:
@yanesca As discussed offline, I have changed the implementation of the internal process functions, to return the unsupported error.
Once reviewed and approved, I will rebase to resolve the conflicts

RonEld commented:
I have taken Andres' stress test from #5079 and modified to test cloning:
~~~
int sha256()
{
    int ret, next;
    size_t i, j;
    mbedtls_sha256_context *ctx;
    mbedtls_sha256_context *cloned_ctx;
    unsigned char output[32];
    size_t proc_len[SHA256_CTX_COUNT];
    size_t next_len;

    if( ( ctx = (mbedtls_sha256_context *)mbedtls_calloc( SHA256_CTX_COUNT,
                                    sizeof( mbedtls_sha256_context ) ) ) == NULL )
    {
        mbedtls_printf( "  !  mbedtls_calloc() return NULL\r\n" );
        return( -1 );
    }
    
    if( ( cloned_ctx = (mbedtls_sha256_context *)mbedtls_calloc( SHA256_CTX_COUNT,
                                    sizeof( mbedtls_sha256_context ) ) ) == NULL )
    {
        mbedtls_printf( "  !  mbedtls_calloc() return NULL\r\n" );
        return( -1 );
    }

    /* Initialise all context structures */
    for( i = 0; i < SHA256_CTX_COUNT; i++ )
    {
        mbedtls_sha256_init( &ctx[i] );
        mbedtls_sha256_init( &cloned_ctx[i] );
    }

    for( j = 1; j < REPETITIONS; j++ )
    {
        int cloned[SHA256_CTX_COUNT] = {0};
        /* Start the contexts */
        for( i = 0; i < SHA256_CTX_COUNT; i++ )
            mbedtls_sha256_starts( &ctx[i], 0 );

        memset( proc_len, 0, sizeof( proc_len ) );

        do
        {
            next = 0;

            for( i = 0; i < SHA256_CTX_COUNT; i++ )
            {
                /* Nothing left to do */
                if( proc_len[i] == sha256_ilens[i] )
                    continue;

                next = 1;

                next_len = sha256_ilens[i] / SPLIT + 1;
                next_len = ( next_len + proc_len[i] <= sha256_ilens[i] ) ? next_len :
                            sha256_ilens[i] - proc_len[i];
                if( proc_len[i] >= sha256_ilens[i] / 2 && !cloned[i] )
                {
                    mbedtls_sha256_clone( &cloned_ctx[i], &ctx[i] );
                    cloned[i] = 1;
                }
                mbedtls_sha256_update( ( cloned[i])? &cloned_ctx[i]:&ctx[i], sha256_inputs[i] + proc_len[i],
                                    next_len );
                    
                proc_len[i] += next_len;
            }
        }
        while( next == 1 );

        /* Verify the outputs */
        for( i = 0; i < SHA256_CTX_COUNT; i++ )
        {
            mbedtls_sha256_finish( cloned[i]? &cloned_ctx[i]:&ctx[i], output );

            if( memcmp( output, sha256_results[i], sizeof( output ) ) != 0 )
            {
                mbedtls_printf( "%d:  !  memcmp() result is not 0 at "
                                " repetition %u:%u\r\n",
                                __LINE__, j, i );
                ret = -2;
                goto exit;
            }
        }
    }

    ret = 0;

exit:
    for( i = 0; i < SHA256_CTX_COUNT; i++ )
        mbedtls_sha256_free( &ctx[i] );

    mbedtls_free( ctx );

    return( ret );
}
~~~
This test passes with the original implementation. So, I will remove the original commit, keeping the style fix, and latest commit for not supporting the internal process function

RonEld commented:
@yanesca As mentioned in previous comment, I have rebased to resolved conflicts, and removed the first commit that is not needed.
A reference branch from before the rebase is in https://github.com/RonEld/mbed-os/tree/cryptocell_sha_alt_fixes_before_rebase

cmonr commented:
@yanesca Please re-review.
@ARMmbed/mbed-os-tls Please review.

yanesca commented:
Looks good to me! I will approve formally as soon as @ARMmbed/mbed-os-tls has approved it too.

RonEld commented:
@yanesca I don't think this should be reviewed by @ARMmbed/mbed-os-tls but by @ARMmbed/mbed-os-crypto as this is a crypto bug fix

dgreen-arm reviewed (approved):
Looks good to me.

RonEld commented:
@cmonr @0xc0170 Could you start CI please?

cmonr commented:
Will start CI when able.

We're wrapping up RC3 PRs right now.

cmonr commented:
CI started
